rc/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+use std::ops::Deref;
+use std::ptr;
+
+pub struct MyRc<T> {
+  data: *mut T,
+  count: *mut usize,
+}
+
+impl<T> MyRc<T> {
+  pub fn new(t: T) -> MyRc<T> {
+      let data_box  = Box::new(t);
+      let count_box = Box::new(1);  // one reference to the underlying data
+      let data_ptr  = Box::into_raw(data_box);  // now we own the heap data
+      let count_ptr = Box::into_raw(count_box);
+      MyRc{data: data_ptr, count: count_ptr}
+  }
+
+  pub fn consume(self) -> Result<T, MyRc<T> > {
+    unsafe {
+      if *self.count == 1 {
+        let t: T = ptr::read(self.data);
+        // This will prevent us from dropping the data twice
```

Comment:
Unfortunately it also causes you to not free the memory which holds the data (nor the memory which holds the reference count). Using valgrind (as described on Piazza) reveals this bug.

It might be a good idea to do you own cleanup here, including freeing those allocations, and then forget `self`.

Alternatively, you can continue to use the destructor, but work out some invariant that lets the destructor know when it should both free the allocation and drop the contents vs. just free the allocation.